src/ert/ensemble_evaluator/builder/_ensemble_builder.py

```diff
@@ -217,3 +223,6 @@ def build(self) -> _Ensemble:
                 retry_delay=self._retry_delay,
                 custom_port_range=self._custom_port_range,
             )
+        if self._id is not None:
```

Comment:
Can you just use the same as before, but using the `_LegacyEnsemble(reals, self._metadata, *self._legacy_dependencies).set_id(self.id_)` in both returns?

Reply:
The default value of `id` is `None`. I was uncertain if there should be a default `uuid` set instead?

As the implementation of setting an `id` enforces the caller to only set it once, I wondered if it would be too strict to force them to include it in the builder. This way the `id` can be included in the builder or be added post build.

I am very open to suggestions here.. Should we fail if the id is not set in the builder?

Reply:
> Should we fail if the id is not set in the builder?` 

I think the `id` should be set only in the builder, what I mean is why to set it post build? Therefore my opinion would be like "Yes, we should fail." :)

Reply:
So then it should not have a default either, and fail if the caller has not explicitly set the id on the builder before calling `build`?

Reply:
Currently it's set when constructing `EnsembleEvaluator` https://github.com/equinor/ert/blob/1764672345ceddcf75b688b097ee5f5572384ffc/src/ert/shared/models/base_run_model.py#L315, we might just inject it as a parameter for `_Ensemble` too or what do you think?

Reply:
Am I understanding correctly if I suggest the following?

```diff
diff --git a/src/ert/shared/models/base_run_model.py b/src/ert/shared/models/base_run_model.py
index a59d715d3..2cd67a132 100644
--- a/src/ert/shared/models/base_run_model.py
+++ b/src/ert/shared/models/base_run_model.py
@@ -294,13 +294,17 @@ class BaseRunModel:
     def run_ensemble_evaluator(
         self, run_context: RunContext, ee_config: EvaluatorServerConfig
     ) -> int:
-        ensemble = EnsembleBuilder.from_legacy(
+        ensemble = (EnsembleBuilder.from_legacy(
             run_context,
             self.get_forward_model(),
             self._queue_config,
             self.ert().analysisConfig(),
             self.ert().resConfig(),
-        ).build()
+        )
+        .self.set_id(str(uuid.uuid1()).split("-", maxsplit=1)[0])
+        .build()
+        )
+
 
         self.ert().initRun(run_context)
 
@@ -308,7 +312,6 @@ class BaseRunModel:
             ensemble,
             ee_config,
             run_context.iteration,
-            ee_id=str(uuid.uuid1()).split("-", maxsplit=1)[0],
         ).run_and_get_successful_realizations()
 
         self.deactivate_failed_jobs(run_context)
```

Reply:
Hm, that's also an option I think it's fine :)

Reply:
Alright, and given:

>  Therefore my opinion would be like "Yes, we should fail." :)

There will then be raised an exception if anyone tries to execute `build` without setting the id first.

Reply:
> Furthermore, being consistent with how the Experiment protocol is used in e.g. BaseRunModel, I do not include the id as part of the __init__. Instead the id is set after an Ensemble is created (but this will also be done within the builder).

So, I decided to disagree with myself.. I don't like the pattern of creating an instance of an object and then remembering to always set the property afterwards. I think the `id` should be required on init instead. The builder will work as proposed above though!

I also think we should change this in any `Experiment` classes.

Reply:
I've included the changes and pushed it now, please let me know if you agree on how it looks :)